csv/csv.go

```diff
@@ -0,0 +1,131 @@
+package csv
+
+import (
+	"bytes"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/gobuffalo/pop"
+	"github.com/pkg/errors"
+)
+
+// ErrInvalidCSV is returned when the input CSV file is not valid.
+var ErrInvalidCSV = errors.New("invalid CSV file")
+
+// Importer defines the options for the pop CSV import feature.
+type Importer struct {
+	cnx     *pop.Connection
+	elapsed int64
+	// Comma is the field delimiter.
+	// It is set to comma (',') by default.
+	// Comma must be a valid rune and must not be \r, \n,
+	// or the Unicode replacement character (0xFFFD).
+	Comma rune
+}
+
+func (ci *Importer) timeFunc(name string, fn func() error) error {
+	now := time.Now()
+	err := fn()
+	atomic.AddInt64(&ci.elapsed, int64(time.Now().Sub(now)))
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+// FromReader loads a CSV file from an io.Reader, and insert rows into the datasource
+// using the given connection. It allows to use a custom file reader, such as a remote
+// one.
+//
+// Example usage:
+// 	imp := csv.NewImporter(tx)
+//  	fd, err := os.Open("./csv/files/my_table_data.csv")
+//	if err != nil {
+//		return err
+//	}
+//	defer fd.Close()
+//	return ci.FromReader(fd, "my_table")
+func (ci *Importer) FromReader(fd io.Reader, model interface{}) error {
+	m := &pop.Model{Value: model}
+	return ci.timeFunc("CSV import", func() error {
+		r := csv.NewReader(fd)
+		r.Comma = ci.Comma
+		records, err := r.ReadAll()
+		if err != nil {
+			return err
+		}
+
+		if len(records) <= 1 {
+			return ErrInvalidCSV
+		}
+
+		// Create the insert query
+
+		// The first row contains the columns
+		var qBuffer bytes.Buffer
+		h := records[0]
+		qBuffer.WriteString(fmt.Sprintf("INSERT INTO %s (", m.TableName()))
+		hl := len(h) - 1
+		for i := 0; i < hl; i++ {
+			qBuffer.WriteString(h[i])
```

Comment:
should we clean up this field name to avoid injection?

Reply:
Agreed.

Reply:
thank you 👍